website/blog/2024-07-26-pyfunc-in-practice/index.md

```diff
@@ -0,0 +1,930 @@
+    ) -> None:
+    """
+    Trains the ensemble of models using the provided preprocessed training data.
+
+    Args:
+        X_train_processed (pd.DataFrame): Preprocessed feature matrix for training.
+        y_train (pd.Series): Target variable for training.
+        save_path (str): Directory path where trained models will be saved.
+    """
+    # Create the directory for saving models if it does not exist
+    os.makedirs(save_path, exist_ok=True)
+
+    # Iterate over each model and its parameter grid
+    for model_name, (model, param_grid) in self.models.items():
+        # Perform GridSearchCV to find the best hyperparameters for the current model
+        grid_search = GridSearchCV(
+            model, param_grid, cv=5, n_jobs=-1, scoring="neg_mean_squared_error"
+        )
+        grid_search.fit(
+            X_train_processed, y_train
+        )  # Fit the model with the training data
+
+        # Save the best estimator from GridSearchCV
+        best_model = grid_search.best_estimator_
+        self.fitted_models[model_name] = best_model
+
+        # Save the trained model to disk
+        joblib.dump(best_model, os.path.join(save_path, f"{model_name}.pkl"))
+```
+
+### Defining a Custom `predict` Method to Aggregate Multi-model Predictions
+
+To streamline the inference process, every PyFunc model should define a custom `predict` method as the single entry point for inference. This approach abstracts the model's internal workings at inference time, whether dealing with a custom PyFunc model or an out-of-the-box MLflow built-in flavor for popular ML frameworks.
+
+The custom `predict` method for the ensemble model is designed to collect and combine predictions from the sub-models, supporting various aggregation strategies (e.g., average, weighted). The process involves the following steps:
+1. Load the sub-model predictions aggregation strategy based on the user-defined approach.
+2. Load the models to be used for inference.
+3. Preprocess the input data.
+4. Collect predictions from individual models.
+5. Aggregate the model predictions according to the specified strategy.
+
+```python
+def predict(self, context, model_input: pd.DataFrame) -> np.ndarray:
+    """
+    Predicts the target variable using the ensemble of models based on the selected strategy.
+
+    Args:
+        context: MLflow context object.
+        model_input (pd.DataFrame): Input features for prediction.
+
+    Returns:
+        np.ndarray: Array of predicted values.
+
+    Raises:
+        ValueError: If the strategy is unknown or no models are fitted.
+    """
+    # Check if the 'strategy' column is present in the input DataFrame
+    if "strategy" in model_input.columns:
+        # Extract the strategy and drop it from the input features
+        print(f"Strategy: {model_input['strategy'].iloc[0]}")
+        strategy = model_input["strategy"].iloc[0]
+        model_input = model_input.drop(columns=["strategy"])
```

Comment:
```suggestion
        model_input.drop(columns=["strategy"], inplace=True)
```

Reply:
Thank you for pointing that out, updated it.